reactive-charming-details.md

```diff
@@ -0,0 +1,233 @@
+run_container function sets the "nginx.started" state this insures the
+container is not started over and over again.
+```python
+@when('nginx.available', 'docker.available')
+@when_not('nginx.started')
+def run_container(webroot=None):
+    '''
+    Wrapper method to launch a docker container under the direction of Juju,
+    and provide feedback/notifications to the end user.
+    '''
+    if not webroot:
+        webroot = config['webroot']
+    # Run the nginx docker container.
+    run_command = [
+        'docker',
+        'run',
+        '--restart',
+        'on-failure',
+        '--name',
+        'docker-nginx',
+        '-v',
+        '{}:/usr/share/nginx/html:ro'.format(webroot),
+        '-p',
+        '{}:80'.format(config['port']),
+        '-d',
+        'nginx'
+    ]
+    check_call(run_command)
+    hookenv.open_port(config['port'])
+    reactive.remove_state('nginx.stopped')
+    reactive.set_state('nginx.started')
+    hookenv.status_set('active', 'Nginx container started')
+```
+
+#### stop_container
+The stop_container function is decorated with
+`@when('nginx.stop', 'docker.available')`.  The "nginx.stop" state is an
+indication to stop the container in which case it will set the "nginx.stopped"
+state. The decorator `@when_not('nginx.stopped')` protects this function from
+being called repeatedly. Note the `reactive.remote_state('nginx.stop')` and
+`reactive.set_state('nginx.stopped')`calls to assert a stopped state.
+```python
+@when('nginx.stop', 'docker.available')
+@when_not('nginx.stopped')
+def stop_container():
+    '''
+    Stop the NGinx application container, remove it, and prepare for launching
+    of another application container so long as all the config values are
+    appropriately set.
+    '''
+    hookenv.status_set('maintenance', 'Stopping Nginx container')
+    # make this cleaner
+    try:
+        check_call(['docker', 'kill', 'docker-nginx'])
+    except:
+        pass
+    try:
+        check_call(['docker', 'rm', 'docker-nginx'])
+    except:
+        pass
+    reactive.remove_state('nginx.started')
+    reactive.remove_state('nginx.stop')
+    reactive.set_state('nginx.stopped')
+    hookenv.status_set('waiting', 'Nginx container stopped')
+```
+
+#### configure_website_port
+The configure_website_port is a function that handles the http relationship,
+specifically the port. This function is decorated with
+`@when('nginx.started', 'website.available')` to only run when Nginx is started
+and the webiste is available.
+
+```python
+@when('nginx.started', 'website.available')
+def configure_website_port(http):
+    '''
+    Relationship context, used in tandem with the http relation stub to provide
+    an ip address (default to private-address) and set the port for the
+    relationship data
+    '''
+    serve_port = config['port']
+    http.configure(port=serve_port)
+    hookenv.status_set('active', '')
+```
+
+## Now write your own compose and reactive charm
+
```

Comment:
You might want to start with this section as coming to this as someone wanting a simple docker charm the framework is less interesting than what they need to do.